packages/noya-state/src/reducers/layerReducer.ts

```diff
@@ -438,50 +439,89 @@ export function layerReducer(
             : draftPage;
 
         draft.selectedLayerIds = [];
+        const parentLayer = Layers.isParentLayer(selectedLayer)
+          ? selectedLayer
+          : (draftPage as Layers.ParentLayer);
 
-        let parentLayer = draftPage as Layers.ParentLayer;
-
-        if (Layers.isParentLayer(selectedLayer)) {
-          parentLayer = selectedLayer;
-        }
+        const meta = draft.sketch.user[draftPage.do_objectID] ?? {
+          zoomValue: 1,
+          scrollOrigin: '{0,0}',
+        };
 
         const viewportCenter = {
           x: context.canvasSize.width / 2,
           y: context.canvasSize.height / 2,
         };
 
-        const meta = draft.sketch.user[draftPage.do_objectID] ?? {
-          zoomValue: 1,
-          scrollOrigin: '{0,0}',
-        };
-
         const parsed = Primitives.parsePoint(meta.scrollOrigin);
+        const currentCanvasFrame = {
+          x: parsed.x,
+          y: parsed.y,
+          width: context.canvasSize.width,
+          height: context.canvasSize.height,
+        };
 
         layers.forEach((layer) => {
+          const layerFrame = layer.frame;
           const { pageIndex, indexPaths } = findPageLayerIndexPaths(
             state,
             (l) => l.do_objectID === layer.do_objectID,
           )[0];
 
           if (layer._class === 'page') return;
-
           const newLayer = produce(copyLayer(layer), (l) => {
             l.name = layer.name;
-            l.frame = {
-              ...l.frame,
-              x: viewportCenter.x - parsed.x - layer.frame.width / 2,
-              y: viewportCenter.y - parsed.y - layer.frame.height / 2,
-            };
           });
 
-          if (!Layers.isPageLayer(parentLayer)) {
-            addToParentLayer(parentLayer.layers, newLayer);
-          } else if (currentPageIndex === pageIndex && indexPaths.length > 0) {
+          if (
+            isLayerVisible(layer, currentCanvasFrame) &&
+            currentPageIndex === pageIndex &&
+            indexPaths.length > 0
+          ) {
+            // If the layer is on the current page and is visible
             addSiblingLayer(draftPage, indexPaths[0], newLayer);
-          } else if (!Layers.isPageLayer(selectedLayer)) {
-            addSiblingLayer(draftPage, selectedLayerIndexPath, newLayer);
+          } else if (!Layers.isPageLayer(parentLayer)) {
+            // If the selected layer is a parent layer but not a page layer
+            newLayer.frame = {
+              ...newLayer.frame,
+              x: parentLayer.frame.width / 2 - layerFrame.width / 2,
```

Comment:
To get a layer midpoint, we can use 

```ts
const parentBounds = getBounds(parentLayer.frame)
parentBounds.midX
parentBounds.midY
```

Reply:
And maybe a function that sets the midpoint of a frame would help?

Untested:

```ts
// noya-geometry rect.ts
export function moveRectCentered<T extends Rect>(rect: T, point: Point): T {
  const bounds = createBounds(rect);

  const delta: Point = {
    x: bounds.midX - point.x,
    y: bounds.midY - point.y,
  };

  return {
    ...rect,
    x: rect.x + delta.x,
    y: rect.y + delta.y,
  };
}
```

Reply:
The parents midx and y are in relationship to the canvas not internal x and y so it doesn't work